### DISCUSSION TITLE
[Feature request] Highlight differences during Conflicted Copy errors

### DISCUSSION THREAD

**Discussion by yingw787** [standardnotes/forum#2602] [Suggestions] [answered]
Hello!

My Internet is extremely spotty sometimes and I've been getting Conflicted Copy errors quite frequently. One thing that might help me resolve Conflicted Copy errors better is highlighting where the conflict occurs. Is there a merge conflict underlying view that can be exposed to dev/canary users by any chance? Sometimes my notes are quite large and finding the differences takes a while and correct resolution may have low confidence; this would be a big help on my end. If you think this is a small issue and you are open to a PR, some pointers to where I could look first would be really helpful too. Thanks for your help, and please let me know if there is more information you need me to submit.

iOS Standard Notes version: 3.0.16
iOS Version: 13.1.3
iPhone Version: iPhone 7 (Model MN8N2LL/A)

Ying

**Comment by atmoio** [comment]
This would be nice to have. Definitely not a small issue though. We'd need client-side JavaScript-based diffing. One place to start is here: https://github.com/standardnotes/web/blob/master/app/assets/javascripts/app/directives/views/conflictResolutionModal.js

This is currently an unused modal. Does not have diffing.

**Comment by wonderfullyrich** [comment]
Wanted to follow up this issue, as I second yingw787 point.  I end up using StandardNotes on a daily basis and am in and out of internet all the time (I work on a ship so internet is not assured) as well as being in developing world countries with internet challenges.  The aggressive syncing that StandardNotes does seems to often create multiple conflicting copies of a note even if I'm just on a single device.  One simple solution in the interim would be just to have an offline mode which the user can enable and stop the attempts at syncing. 

As for when it happens, I find myself looking at the same idea that yingw787 mentioned about highlighting the differences during a merge.  I can manually do the same on a linux command line with diff, but it would be very nice to see it in the same way notepad++ or bbedit does it.

iOS Standard Notes 3.0.21
Android Standard Notes 3.0.19
MacOS Standard Notes 3.3.3

Rich

**Comment by arielsvg** [comment]
Thanks for sharing your experience, it's really eye-opening! There definitely are ways we could make syncing more robust (syncing is tricky in general) and we'll keep that in mind…

**Comment by denmon99** [comment]
> This would be nice to have. Definitely not a small issue though. We'd need client-side JavaScript-based diffing. One place to start is here: https://github.com/standardnotes/web/blob/master/app/assets/javascripts/app/directives/views/conflictResolutionModal.js
> 
> This is currently an unused modal. Does not have diffing.

When I have a conflict I go to my desktop app, and copy/paste the two conflicting notes into Notepad++ on Windows and run its Compare plugin. Other tools like WinMerge would work as well. 

A simpler alternative to implementing the whole diffing process in-app would be to have a way to launch an external diff tool.  I can imagine being able to right-click one of the conflicting notes, choose "Launch external diff", and have it write out temp files of the the two conflicting note contents and launch WinMerge or similar on those temp files. 

This doesn't solve getting the modified content back into Standard Notes, but for me usually there are only a few small changes between them and I can tell which note to keep or what line to move between them.

**Comment by virtuallyvlad** [comment]
> [...] usually there are only a few small changes between them and I can tell which note to keep or what line to move between them.

FYI as of now SN also [gets rid of all of the history on one of the copies](https://github.com/standardnotes/forum/issues/2290), I believe typically the newer one, so it's usually better to keep the old one and just merge in the changes, otherwise you lose all the history on your note.

**Comment by denmon99** [comment]
That's an important distinction, thanks - I was unaware. I will definitely have to watch for that!

**Comment by hamoid** [comment]
Hi. I recently installed SN. I'm liking everything except the conflict resolution side of things. I often have conflicts because I work on multiple devices which are not always online. 

I find resolving the conflicts very hard. I would like to copy some lines from A to B and other lines from B to A, and I don't see a way to do that. I'm used to tools like [Meld](https://meldmerge.org/) or https://www.scootersoftware.com/

I could use @denmon99 's approach when really needed, but not regularly.

Is there any plan to add some kind of merging of changes between two files?

**Reply by hamoid** [reply]
Or at least making it possible to edit the text in the conflict view, and highlighting the differences?

**Reply by virtuallyvlad** [reply]
I wouldn't hold my breath for SN implementing change merging; best case scenario you could create a feature request for editing text in conflict view and they could conceivably add that.

If you have to use SN on a spotty connection, I would just make sure that device syncs to the cloud before starting to edit on another device.  Or if you cannot sync at all at least avoid making edits to the same note (e.g., create new notes instead).

For another alternative, you could check out Apple Notes as I found it dealt with conflicting copies very well (versus SN/Obsidian/Logseq, which in my testing did not).

**Comment by virtuallyvlad** [comment, marked as answer]
I think this can now be closed as we have diff highlighting.

![image](https://github.com/user-attachments/assets/3e219043-93ff-486d-a089-8bb4511e93bc)
